BPF: Generate locked insn for __sync_fetch_and_add() with cpu v1/v2

yonghong-song opened the pull request:
There are two commits in this pull request:
  - first commit is to revert https://github.com/llvm/llvm-project/pull/101428
  - second commit removed `atomic_fetch_and_*()` to `atomic_<op>()` conversion if return value is not used, but preserving `__sync_fetch_and_add()` to locked insn with cpu v1/v2.

Commit:
Revert "BPF: Ensure __sync_fetch_and_add() always generate atomic_fetch_add insn (#101428)"

This reverts commit c566769d7c097e3956c6b36c2040bd8baa2e9929.

See discussion in [1]. Currently, with -mcpu=v1/v2, atomic_fetch_add() insn is generated
for (void)__sync_fetch_and_add(...). This breaks backward compatibility since
there are codes who runs on old system (< 5.12) which does not support atomci_fetch_add().

Now let revert previous comment ([1]) and add additional logic in the next patch
to ensure for (void)__sync_fetch_and_add(...),
  v1/v2:  generate locked add insn
  >= v3:  generate atomic_fetch_add insn

  [1] https://github.com/llvm/llvm-project/pull/101428

Commit:
[BPF] Do not convert atomic_fetch_and_*() to atomic_<op>()'s

In previous commit, atomic_fetch_and_*() operations are converted
to atomic_<op>()'s if there are no return values. This is not
what we want, we would like to preserve atomic_fetch_and_*() insn
so bpf jit can add proper barrier insns.

Preserving atomic_fetch_and_*() are okay for most __sync_fetch_and_*()
functions, but not for __sync_fetch_and_add() since
__sync_fetch_and_add() has been used to generic locked insns
in cpu=v1/v2.

So after preserving atomic_fetch_and_*() even if return value
is not used, XFADDD in BPFInstrInfo.td is adjusted to emit
locked add insn for cpu v1/v2 and to emit atomic_fetch_and_add()
for cpu >= v3.

yonghong-song commented:
cc @brycekahle

4ast commented on llvm/test/CodeGen/BPF/atomics.ll:
```diff
@@ -1,19 +1,22 @@
-; RUN: llc < %s -march=bpfel -verify-machineinstrs -show-mc-encoding | FileCheck --check-prefixes=CHECK,CHECK-V2 %s
-; RUN: llc < %s -march=bpfel -verify-machineinstrs -show-mc-encoding -mcpu=v3 | FileCheck --check-prefixes=CHECK,CHECK-V3 %s
+; RUN: llc < %s -march=bpfel -verify-machineinstrs -show-mc-encoding | FileCheck %s
```
should we add -mcpu=v1 here and maybe flip a default to -mcpu=v3 ?
So that by default memory ordering is preserved and users that want to see xadd insn for (void)fetch_and_add
have to explicitly request it with -mcpu=v1.

yonghong-song replied:
We can do this. This may cause verification failure or other things from v1 to v3. I guess we can have a separate patch just to flip the default. I can then post it to bpf mailing list to let people know.

Regarding to selftests, current in llvm/test/CodeGen/BPF directory (not including sub-directory), we only '-mcpu=v1' for llvm-objdump since llvm-objdump has been improved to dump insn based on cpu v4.
Otherwise, -mcpu=v1 is the default and not specified in the test.
```
[~/work/llvm-project/llvm/test/CodeGen/BPF (main)]$ grep "mcpu=v1" *.ll
objdump_cond_op.ll:; RUN: llc -mtriple=bpfel -filetype=obj -o - %s | llvm-objdump --no-print-imm-hex --mcpu=v1 -d - | FileCheck %s
objdump_static_var.ll:; RUN: llc -mtriple=bpfel -filetype=obj -o - %s | llvm-objdump --no-print-imm-hex --mcpu=v1 -d - | FileCheck --check-prefix=CHECK %s
objdump_static_var.ll:; RUN: llc -mtriple=bpfeb -filetype=obj -o - %s | llvm-objdump --no-print-imm-hex --mcpu=v1 -d - | FileCheck --check-prefix=CHECK %s
[~/work/llvm-project/llvm/test/CodeGen/BPF (main)]$
```

Later when we want to make cpu v3 as the default, all related tests can be modified at that time.

eddyz87 commented on llvm/lib/Target/BPF/BPFInstrInfo.td:
```diff
@@ -806,6 +806,7 @@ class XADD<BPFWidthModifer SizeOp, string OpcodeStr, PatFrag OpNode>
 let Constraints = "$dst = $val" in {
   let Predicates = [BPFNoALU32] in {
     def XADDW : XADD<BPF_W, "u32", atomic_load_add_i32>;
+    def XADDD : XADD<BPF_DW, "u64", atomic_load_add_i64>;
```
Would it be possible to get rid of `XADD` in favour of `ATOMIC[32]_NOFETCH` and a few `def : Pat`? I tried to do so [remove-xadd.txt](https://github.com/user-attachments/files/16807144/remove-xadd.txt), but that is not enough.

yonghong-song replied:
Not sure about this. Let me give a try.

yonghong-song replied:
@eddyz87 I checked your code. It won't work because with BPFNoALU32 does not allow sub-registers. In current implementation, 64bit atomic_nofetch insns are available at cpu v1, but 32bit atomic nofetch (except xadd) are not available at cpu v1. so 32bit xadd needs specialization anyway. so your approach will work for 64bit xaddd, but not for xadd. So I suggest to use the current implementation. Let me know what you think.

eddyz87 replied:
Right, that thing was not complete and I couldn't figure why, thank you for explaining.
Here is a fixed version:  [remove-xadd.txt](https://github.com/user-attachments/files/16816421/remove-xadd.txt)
What I like here is that all xadd variants are grouped together, but up to you.

yonghong-song replied:
@eddyz87 What you suggested indeed simpler and I just incorporated it into the latest revision. I made two changes:
  - Remove 'let mayStore = 1' in ATOMIC_NOFETCH. I am not whether this is needed or not. Previous class does not have this.
  - Remove 'MI.getOpcode() != BPF::XADDW32' in BPFMIChecking.cpp. XADDW32 is only available with alu32, in which case return value is okay.

Please help take a look again in case that I missed anything. Thanks!

eddyz87 replied:
> Remove 'let mayStore = 1' in ATOMIC_NOFETCH. I am not whether this is needed or not. Previous class does not have this.

I think tablegen wanted me to add this flag when I added patterns. If it does not report warnings on you side, then ok.

> Remove 'MI.getOpcode() != BPF::XADDW32' in BPFMIChecking.cpp. XADDW32 is only available with alu32, in which case return value is okay.

`XADDW32` is not a member of any pattern, so I agree, this should be fine.

yonghong-song replied:
For
>  I think tablegen wanted me to add this flag when I added patterns. If it does not report warnings on your side, then ok.

No, I didn't see any warnings.

eddyz87 replied:
Ok, all good then

eddyz87 commented on llvm/lib/Target/BPF/BPFMIChecking.cpp:
```diff
@@ -171,81 +168,14 @@ bool BPFMIPreEmitChecking::processAtomicInsts() {
       }
     }
   }
-
```
After these changes, for atomic fetch and add we would have  `atomic_fetch_add` for cpu v3/v4 and `lock *(u64 *)...` for cpu v1/v2. However, for and/or/xor `atomic_fetch_*` versions would be generated. Why only add gets special treatment?
E.g.:
```
$ cat test3.c 
unsigned long xadd(unsigned long *p, unsigned long v)
{
  __sync_fetch_and_add(p, v);
  __sync_fetch_and_and(p, v);
  return 0;
}

$ clang -O2 --target=bpf -mcpu=v1 -S test3.c -o -
...
xadd:                                   # @xadd
# %bb.0:                                # %entry
	r3 = r2
	lock *(u64 *)(r1 + 0) += r3
	r2 = atomic_fetch_and((u64 *)(r1 + 0), r2)
	r0 = 0
	exit
...
```

yonghong-song replied:
I am aware of this issue and I debate with myself. atomic-* insns are introduced and supposed to be in v3. But somehow we allowed __sync_fetch_and_and() style functions exist for v1 as well for quite some time. In previous implementation (e.g. llvm19), if no return value, the atomic_fetch_and() will become locked and insn. Now to keep __sync_fetch_and_and() barrier semantic, we cannot change to locked insn any more.

Should we just ensure all atomic_fetch_*() insns (except atomic_fetch_add()) only supports cpu=v3? This is my original thinking as well, but I am afraid some people may already use it (for >= 5.12 kernel) for v1 . If we disable usage in v1 then some existing use case may break.

Or we assume people only use locked add insn for v1/v2, all other atomic insns should be only available >= v3?

@4ast Any comments?

4ast replied:
> if no return value, the atomic_fetch_and() will become locked and insn. 

There is no "locked and" prior to v3.
atomic_fetch_and() should always generate atomic_fetch_and().

atomic_fetch_ADD() is the only exception, because we had xadd insn from the beginning.

> Should we just ensure all atomic_fetch_*() insns (except atomic_fetch_add()) only supports cpu=v3?

I don't think we need to add such restriction.
If code has  atomic_fetch_XX() it should stay atomic_fetch_XX insn regardless of v1,v2,v3.
The only exception is XX==ADD, because we had this special xadd insn early on and had
this buggy atomic_fetch_add -> xadd code generation that we have to preserve for backward compat reasons.

Even with -mcpu=v1 it's fine to translate sync_fetch_and_xor() to atomic_fetch_xor insn.
We do that for may_goto insn too. -mcpu=vX makes a difference when compiler has a choice.
For v1 it can generate one sequence while for v3 another sequence of insns.
For may_goto and for sync_fetch_and_xor there is no choice. Either generate that insn or error.
It's better to generate working code than error. -mcpu is not a binding contract.

yonghong-song replied:
Okay, sounds good to me. Indeed, for non-XADD* insns, which are introduced in llvm12, regardless whether they use atomic_fetch_*() or atomic_<op>(), they are all introduced in 5.12 kernel and llvm12. So from source code perspective, they should be okay whether underlying we use atomic_fetch_*() or atomic_<op>() or not, and also regardless of whether it is v1 or v3. Maybe somebody already started to use it with v1...

So I will keep current non-XADD* behavior in the patch.

eddyz87 replied:
Understood, thank you for explaining

Commit:
Simplify XADD/XADDD pattern matching

The patch is from Eduard.

class XADD and class ATOMIC_NOFETCH are almost the same except
the pattern matching string. Removing class XADD and defining
two Pat's (pattern -> insn) for 32bit and 64bit XADD/XADDD
to use ATOMIC_NOFETACH can simplify code a lot.

th0rex commented:
Hi, this commit broke some of our code, I am not sure whether this is intentional or not. It seems to me that anything using the return value of e.g. `__sync_fetch_and_add` doesn't work anymore with `-mcpu=v2` (nor with `-mcpu=v1`).

To reproduce:

On the parent of this commit (57fe53cae40351ebd079a9a0105addf4ad2e97dd) the following test is working:

```
~/llvm-project ((57fe53ca))> git status
HEAD detached at 57fe53cae403
Changes not staged for commit:
  (use "git add/rm <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
        modified:   llvm/test/CodeGen/BPF/atomics.ll

no changes added to commit (use "git add" and/or "git commit -a")

~/llvm-project ((57fe53ca))> git diff
diff --git a/llvm/test/CodeGen/BPF/atomics.ll b/llvm/test/CodeGen/BPF/atomics.ll
index 0c16c49f2a87..5e71dcef85a2 100644
--- a/llvm/test/CodeGen/BPF/atomics.ll
+++ b/llvm/test/CodeGen/BPF/atomics.ll
@@ -19,3 +19,12 @@ entry:
   atomicrmw add ptr %p, i64 %v seq_cst
   ret void
 }
+
+; CHECK-LABEL: test_load_add_ret_64
+; CHECK: r2 = atomic_fetch_add((u64 *)(r1 + 0), r2)
+; CHECK: encoding: [0xdb,0x21,0x00,0x00,0x01,0x00,0x00,0x00]
+define i64 @test_load_add_ret_64(ptr %p, i64 zeroext %v) {
+entry:
+  %0 = atomicrmw add ptr %p, i64 %v seq_cst
+  ret i64 %0
+}

~/llvm-project ((57fe53ca))> ./build/bin/llc -march=bpfel -mcpu=v2 llvm/test/CodeGen/BPF/atomics.ll
~/llvm-project ((57fe53ca))> # runs successfully
```

With this commit (06c531e808ceeafdf996867a2e8e66960ae4774e), the same test fails:

```
~/llvm-project ((06c531e8))> git status
HEAD detached at 06c531e808ce
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
        modified:   llvm/test/CodeGen/BPF/atomics.ll

no changes added to commit (use "git add" and/or "git commit -a")

~/llvm-project ((06c531e8))> git diff
diff --git a/llvm/test/CodeGen/BPF/atomics.ll b/llvm/test/CodeGen/BPF/atomics.ll
index c17b94af5f7b..fbccfe54135e 100644
--- a/llvm/test/CodeGen/BPF/atomics.ll
+++ b/llvm/test/CodeGen/BPF/atomics.ll
@@ -22,3 +22,12 @@ entry:
   atomicrmw add ptr %p, i64 %v seq_cst
   ret void
 }
+
+; CHECK-LABEL: test_load_add_ret_64
+; CHECK: r2 = atomic_fetch_add((u64 *)(r1 + 0), r2)
+; CHECK: encoding: [0xdb,0x21,0x00,0x00,0x01,0x00,0x00,0x00]
+define i64 @test_load_add_ret_64(ptr %p, i64 zeroext %v) {
+entry:
+  %0 = atomicrmw add ptr %p, i64 %v seq_cst
+  ret i64 %0
+}

~/llvm-project ((06c531e8))> ./build/bin/llc -march=bpfel -mcpu=v2 llvm/test/CodeGen/BPF/atomics.ll
error: <unknown>:0:0: in function test_load_add_ret_64 i64 (ptr, i64): Invalid usage of the XADD return value
```

Like I said, I'm not sure if this is intentional or not - we just switched to inline assembly and call `atomic_fetch_add` directly rather than relying on `__sync_fetch_and_add` which still works as expected, but I wanted to bring this to your attention in case this is not intentional.

yonghong-song commented:
@th0rex thanks for the reporting. The newer behavior in llvm20 is correct. The old one has some issues and some instructions supposed not support in v1/v2 but somehow we did it and their usage is not intentional. But unfortunately there is no way to do backport. See below for some details.

```
$ cat t.c                                                                                                                         
int f1(int *i) {                                                                                                                                             
   return __sync_fetch_and_add(i, 10);                                                                                                                       
}
$ ~/work/llvm-project/llvm/build.15/install/bin/clang --version
clang version 15.0.7 (https://github.com/llvm/llvm-project.git 8dfdcc7b7bf66834a761bd8de445840ef68e4d1a)
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /home/yhs/work/llvm-project/llvm/build.15/install/bin
$ ~/work/llvm-project/llvm/build.15/install/bin/clang --target=bpf -O2 -mcpu=v1 -c t.c                                 
fatal error: error in backend: Invalid usage of the XADD return value
```
You can see a compiler error will be emitted for cpu v1 if the return value is used. The same will be for v2. For v3, it will be okay since v3 enables 32-bit subregister and we allows new insn atomic_fetch_add. for cpu v1/v2, the atomic_fetch_add is not available.

But for 64bit __sync_fetch_and_add(), it seems going through.
```
$ cat t1.c
long f1(long *i) {
   return __sync_fetch_and_add(i, 10);
}
$ ~/work/llvm-project/llvm/build.15/install/bin/clang --target=bpf -O2 -mcpu=v1 -c t1.c
$ ~/work/llvm-project/llvm/build.15/install/bin/llvm-objdump -d t1.o

t1.o:   file format elf64-bpf

Disassembly of section .text:

0000000000000000 <f1>:
       0:       b7 00 00 00 0a 00 00 00 r0 = 10
       1:       db 01 00 00 01 00 00 00 r0 = atomic_fetch_add((u64 *)(r1 + 0), r0)
       2:       95 00 00 00 00 00 00 00 exit
```

So there is a discrepancy between 32bit and 64bit. But intention is for <= llvm19, for v1/v2, no return value should be used (using locked insn) and compiler should issue an error when it does use return value. For v3, it is okay to use return value (using atomic_fetch_add insn).

llvm20 fixed this issue.
```
$ cat t.c
int f1(int *i) {
   return __sync_fetch_and_add(i, 10);
}
[yhs@devbig309.ftw3 ~/tmp4]$ cat t1.c
long f1(long *i) {
  return __sync_fetch_and_add(i, 10);
}
$ ~/work/llvm-project/llvm/build.20/install/bin/clang --target=bpf -O2 -mcpu=v1 -c t.c
t.c:1:5: error: Invalid usage of the XADD return value
    1 | int f1(int *i) {
      |     ^
1 error generated.
$ ~/work/llvm-project/llvm/build.20/install/bin/clang --target=bpf -O2 -mcpu=v1 -c t1.c
t1.c:1:6: error: Invalid usage of the XADD return value
    1 | long f1(long *i) {
      |      ^
1 error generated.
$ ~/work/llvm-project/llvm/build.20/install/bin/clang --target=bpf -O2 -mcpu=v3 -c t.c
$ ~/work/llvm-project/llvm/build.20/install/bin/clang --target=bpf -O2 -mcpu=v3 -c t1.c
$
```

So for 64bit insn, using inline asm with atomc_fetch_add() is the right choice if you intends to use cpu v1/v2.

th0rex commented:
Thanks for the detailed explanation! This makes sense, we'll keep using inline assembly with `atomic_fetch_add` then.
